app/assets/javascripts/pretty-text/addon/engines/discourse-markdown-it.js

```diff
@@ -375,32 +383,111 @@ export function setup(opts, siteSettings, state) {
   opts.discourse.limitedSiteSettings = {
     secureMedia: siteSettings.secure_media,
     enableDiffhtmlPreview: siteSettings.enable_diffhtml_preview,
+    traditionalMarkdownLinebreaks: siteSettings.traditional_markdown_linebreaks,
+    enableMarkdownLinkify: siteSettings.enable_markdown_linkify,
+    enableMarkdownTypographer: siteSettings.enable_markdown_typographer,
+    markdownTypographerQuotationMarks:
+      siteSettings.markdown_typographer_quotation_marks,
+    markdownLinkifyTlds: siteSettings.markdown_linkify_tlds,
   };
 
   const markdownitOpts = {
     discourse: opts.discourse,
     html: true,
-    breaks: !siteSettings.traditional_markdown_linebreaks,
+    breaks: !opts.discourse.limitedSiteSettings.traditionalMarkdownLinebreaks,
     xhtmlOut: false,
-    linkify: siteSettings.enable_markdown_linkify,
-    typographer: siteSettings.enable_markdown_typographer,
+    linkify: opts.discourse.limitedSiteSettings.enableMarkdownLinkify,
+    typographer: opts.discourse.limitedSiteSettings.enableMarkdownTypographer,
   };
 
   if (opts.discourse.markdownItRules !== undefined) {
-    opts.engine = window
-      .markdownit("zero", markdownitOpts) // Preset for "zero", https://github.com/markdown-it/markdown-it/blob/master/lib/presets/zero.js
-      .enable(opts.discourse.markdownItRules);
+    opts.engine = zeroRuleMarkdownEngine(
+      markdownitOpts,
+      opts.discourse.markdownItRules
+    );
   } else {
     opts.engine = window.markdownit(markdownitOpts);
   }
 
-  const quotation_marks = siteSettings.markdown_typographer_quotation_marks;
+  opts.pluginCallbacks = pluginCallbacks;
+  opts.allowListed = allowListed;
+
+  setupMarkdownEngine(opts, opts.discourse.features);
+
+  customMarkdownEngineCallbacks.forEach(([, callback]) => {
+    callback((engineOpts, afterBuild) =>
+      afterBuild(buildCustomMarkdownEngine(engineOpts, opts))
+    );
+  });
```

Comment:
I'm not sure I understand this part of the API. Why do we need to call the callback with a function instead of just making the `buildCustomMarkdownEnginer` function available to the callback?

Reply:
Discussed this in chat, it's because I want only a cook function to be available to the consumer, not the whole engine building pipeline; they don't need the latter.